aten/src/ATen/native/cpu/SpmmReduceKernel.cpp

```diff
@@ -0,0 +1,601 @@
+      row_start = csr_data[m];
+      row_end = csr_data[m + 1];
+
+      scalar_t* out_ptr = out_data + m * K;
+
+      constexpr int64_t kVecSize = Vec::size();
+      constexpr int64_t kVLEN = kVecSize * 4;
+      constexpr int64_t CHUNK_SIZE = 16;
+
+      // reinit the output row for reduce type 'max' and 'min'
+      int64_t count = row_end - row_start;
+      if (count != 0) {
+        Reducer<scalar_t, reduce>::init(out_ptr, K);
+      }
+
+      // blocking on rowwise to reduce write memory bandwidth
+      for (int64_t e0 = row_start; e0 < row_end; e0 += CHUNK_SIZE) {
+        int64_t e1 = std::min(e0 + CHUNK_SIZE, row_end);
+
+        int64_t k = 0;
+        for (; k < K - (K % kVLEN); k += kVLEN) {
+          Vec out_vec0 = Vec::loadu(out_ptr + k);
+          Vec out_vec1 = Vec::loadu(out_ptr + k + kVecSize);
+          Vec out_vec2 = Vec::loadu(out_ptr + k + kVecSize * 2);
+          Vec out_vec3 = Vec::loadu(out_ptr + k + kVecSize * 3);
+          for (const auto e : c10::irange(e0, e1)) {
+            c = col_data[e];
+            scalar_t val = val_data[e];
+            scalar_t* weight_ptr = weight_data + c * K + k;
+
+            Reducer<scalar_t, reduce>::update(out_vec0, Vec::loadu(weight_ptr) * Vec(val));
+            Reducer<scalar_t, reduce>::update(out_vec1, Vec::loadu(weight_ptr + kVecSize) * Vec(val));
+            Reducer<scalar_t, reduce>::update(out_vec2, Vec::loadu(weight_ptr + kVecSize * 2) * Vec(val));
+            Reducer<scalar_t, reduce>::update(out_vec3, Vec::loadu(weight_ptr + kVecSize * 3) * Vec(val));
+          }
+          out_vec0.store(out_ptr + k);
+          out_vec1.store(out_ptr + k + kVecSize);
+          out_vec2.store(out_ptr + k + kVecSize * 2);
+          out_vec3.store(out_ptr + k + kVecSize * 3);
+        }
+        for (; k < K - (K % Vec::size()); k += Vec::size()) {
+          Vec out_vec = Vec::loadu(out_ptr + k);
+          for (const auto e : c10::irange(e0, e1)) {
+            c = col_data[e];
+            scalar_t val = val_data[e];
+            scalar_t* weight_ptr = weight_data + c * K;
+            Reducer<scalar_t, reduce>::update(out_vec, Vec::loadu(weight_ptr + k) * Vec(val));
+          }
+          out_vec.store(out_ptr + k);
+        }
+        for (; k < K; k++) {
+          scalar_t out_val = out_ptr[k];
+          for (const auto e : c10::irange(e0, e1)) {
+            c = col_data[e];
+            scalar_t val = val_data[e];
+            scalar_t* weight_ptr = weight_data + c * K;
+            Reducer<scalar_t, reduce>::update(out_val, weight_ptr[k] * val);
+          }
+          out_ptr[k] = out_val;
+        }
+      }
+
+      if (reduce == SPMM_MEAN && count != 0) {
+        int64_t k = 0;
+        for (; k < K - (K % Vec::size()); k += Vec::size()) {
+          Vec out_vec = Vec::loadu(out_ptr + k);
+          out_vec /= Vec(count);
+          out_vec.store(out_ptr + k);
+        }
+        for (; k < K; k++) {
+          out_ptr[k] /= count;
+        }
+      }
```

Comment:
I guess it would be more efficient to put this inside the compute loop? The `reduce` is a template arg so checking on it is free anyway.

Reply:
good point!

Reply:
for numerical stability, do div at the very last is more.